File: src/backend/services/UserService.ts
import { UserRepository } from "../repositories/UserRepository";
import { User } from "../database/dbStore";

export class UserService {
  private userRepository = new UserRepository();

  public loginOrCreate(username: string, email: string): User {
    if (!username || !email) {
      throw new Error("Username and email are required.");
    }
    
    let user = this.userRepository.findByEmail(email);
    if (!user) {
      user = {
        id: 0,
        username,
        email,
        createdAt: new Date().toISOString()
      };
      user = this.userRepository.save(user);
    } else {
      // Update username if user already exists but typed a different one
      if (user.username !== username) {
        user.username = username;
        this.userRepository.save(user);
      }
    }
    return user;
  }

  public getUserById(id: number): User | undefined {
    return this.userRepository.findById(id);
  }

  public updateCoursePreferences(
    userId: number,
    courseId: number,
    difficulty: "Beginner" | "Intermediate" | "Advanced",
    preferredAiMode: "Text" | "Voice",
    preferredVoice?: string
  ): User {
    const user = this.userRepository.findById(userId);
    if (!user) {
      throw new Error("User not found.");
    }
    user.coursePreferenceId = courseId;
    user.difficulty = difficulty;
    user.preferredAiMode = preferredAiMode;
    if (preferredVoice) {
      user.preferredVoice = preferredVoice;
    }
    return this.userRepository.save(user);
  }
}

File: converted_code/backend/src/main/java/com/zeriq/ai/service/UserService.java
package com.zeriq.ai.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import com.zeriq.ai.model.User;
import com.zeriq.ai.repository.UserRepository;
import java.util.Optional;

@Service
public class UserService {

    @Autowired
    private UserRepository userRepository;

    public User loginOrCreate(String username, String email) {
        return userRepository.findByEmail(email)
                .orElseGet(() -> {
                    User newUser = new User();
                    newUser.setUsername(username);
                    newUser.setEmail(email);
                    newUser.setDifficulty("Beginner");
                    newUser.setPreferredAiMode("Text");
                    return userRepository.save(newUser);
                });
    }

    public Optional<User> getUserById(Long id) {
        return userRepository.findById(id);
    }

    public User updateCoursePreferences(Long userId, Long courseId, String difficulty, String preferredAiMode) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found with id: " + userId));
        
        user.setCoursePreferenceId(courseId);
        user.setDifficulty(difficulty);
        user.setPreferredAiMode(preferredAiMode);
        return userRepository.save(user);
    }
}
